### PR TITLE
pythonPackages.qutip: init at 4.5.2

### PR DESCRIPTION
**Pull request by mcncm**
* Does not use OpenMP by default
* May be overly conservative in disabling python versions older than 3.5
* Note that build time is a bit long, in part due to tests. It might be
  preferable to disable these in the future.

###### Motivation for this change

This is a fairly popular quantum simulation package that it would be nice to have easy access to.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- ~~[ ] Tested execution of all binary files (usually in `./result/bin/`)~~ (N/A)
- [x] Built the package with `nix-shell -p 'python$X.withPackages (ps: [ ps.qutip ])'`, for `X`={`36`,`37`,`38`}, and imported the module in a repl of the respective Python version.
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/qutip/default.nix
Comment (SuperSandro2000) [inline pkgs/development/python-modules/qutip/default.nix:2]: Only use stdenv and stdenv.lib in all other places.

Comment (mcncm) [reply pkgs/development/python-modules/qutip/default.nix:2]: Hi Sandro,

First-time contributor here. I take that to mean that `lib` is unused? Thanks for pointing that out! I admit to roughly pattern-matching against other pythonPackages, and easily could have done a thing or two without knowing why.

While you're taking a look at this PR, I have two other questions if that's all right:
1. The bot `grahamcofborg-eval` failed with the message, "This PR does not cleanly list package outputs after merging." What does that mean? What can I do to fix it?
2. It appears that this bot also doesn't like the maintainer name (`mcncm`, _i.e._ me) isn't in scope. Should it be an attribute of `maintainers`? Why is this building successfully for me if there's an undefined variable?

Thank you!

Comment (SuperSandro2000) [reply pkgs/development/python-modules/qutip/default.nix:2]: > I take that to mean that lib is unused?

It is not really unused but the one place it is used it can easily be replaced with ``stdenv.lib``. Tbh it is just a nitpick to have less inputs. It does not really change anything functional wise.

> It appears that this bot also doesn't like the maintainer name (mcncm, i.e. me) isn't in scope. Should it be an attribute of maintainers? Why is this building successfully for me if there's an undefined variable?

You need to add yourself to https://github.com/NixOS/nixpkgs/blob/master/maintainers/maintainer-list.nix. The contributing guide requires to do that in a separate commit. The commit message should look like ``maintainers: add name``.

> The bot grahamcofborg-eval failed with the message, "This PR does not cleanly list package outputs after merging." What does that mean? What can I do to fix it?

If you follow the details link you can read the log which complains about the missing maintainer.

```
undefined variable 'mcncm' at 
```

https://gist.github.com/GrahamcOfBorg/02f8ab994b78fc43abd018687bec154c

Comment (mcncm) [reply pkgs/development/python-modules/qutip/default.nix]: Hi Sandro,

Thanks for the elaboration. I've removed the dependence on `lib`, added myself as a maintainer, and modified the derivation slightly to add openmp support to qutip.

File: pkgs/development/python-modules/qutip/default.nix
Comment (SuperSandro2000) [inline pkgs/development/python-modules/qutip/default.nix:42]: Those dependencies are probably only needed in one *buildInputs.

**Comment by SuperSandro2000** [comment]
Result of `nixpkgs-review pr 103412` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>python37Packages.qutip</li>
    <li>python38Packages.qutip</li>
  </ul>
</details>

File: pkgs/development/python-modules/qutip/default.nix
Comment (SuperSandro2000) [inline pkgs/development/python-modules/qutip/default.nix:14]:
```suggestion
{ stdenv
, python
, buildPythonPackage
, pythonOlder
, fetchFromGitHub
, setuptoolsBuildHook
, hypothesis
, numpy
, scipy
, matplotlib
, cython
, openmp
, sysctl
}:
```

File: pkgs/development/python-modules/qutip/default.nix
Comment (SuperSandro2000) [inline pkgs/development/python-modules/qutip/default.nix:34]:
```suggestion

  nativeBuildInputs = [ setuptoolsBuildHook ];

  buildInputs = stdenv.lib.optionals stdenv.isDarwin [ openmp ];

```

File: pkgs/development/python-modules/qutip/default.nix
Comment (SuperSandro2000) [inline pkgs/development/python-modules/qutip/default.nix:42]:
```suggestion
```

File: pkgs/development/python-modules/qutip/default.nix
Comment (SuperSandro2000) [inline pkgs/development/python-modules/qutip/default.nix:50]:
```suggestion

  doCheck = false;
```

Tests fail one darwin and linux.

File: pkgs/development/python-modules/qutip/default.nix
Comment (SuperSandro2000) [inline pkgs/development/python-modules/qutip/default.nix:52]:
```suggestion
  checkInputs = [ hypothesis sysctl ];
```

File: pkgs/top-level/python-packages.nix
Comment (SuperSandro2000) [inline pkgs/top-level/python-packages.nix:6119]:
```suggestion
  qutip = callPackage ../development/python-modules/qutip {
    inherit (pkgs.llvmPackages) openmp;
  };
```

**Comment by SuperSandro2000** [comment]
Result of `nixpkgs-review pr 103412` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>python37Packages.qutip</li>
    <li>python38Packages.qutip</li>
  </ul>
</details>

File: pkgs/development/python-modules/qutip/default.nix
Comment (jonringer) [inline pkgs/development/python-modules/qutip/default.nix:50]: please leave a comment saying why tests werent ran

File: pkgs/development/python-modules/qutip/default.nix
Comment (jonringer) [inline pkgs/development/python-modules/qutip/default.nix:48]:
```suggestion
  setupPyGlobalFlags = "--with-openmp";
```

**Review by jonringer — COMMENTED** [review]
Please follow [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md) and [manual#submitting-changes-making-patches ](https://nixos.org/nixpkgs/manual/#submitting-changes-making-patches) and squash the fix-up commits.

 `git rebase -i` is a powerful command which achieves this, I created a small video demonstrating it's use [here](https://www.youtube.com/watch?v=OR8Q-4PLalc). A more indepth text tutorial can be found [here](https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase)

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

**Comment by jonringer** [comment]
to allow tests:
```
diff --git a/pkgs/development/python-modules/qutip/default.nix b/pkgs/development/python-modules/qutip/default.nix
index 14a13683272..b8a104d8cc7 100644
--- a/pkgs/development/python-modules/qutip/default.nix
+++ b/pkgs/development/python-modules/qutip/default.nix
@@ -43,17 +43,14 @@ buildPythonPackage rec {

   enableParallelBuilding = true;

-  buildPhase = ''
-  ${python.interpreter} setup.py --with-openmp bdist_wheel
-  '';
-
-  doCheck = false;
+  setupPyBuildFlags = "--with-openmp";

   checkInputs = [ hypothesis ];

   checkPhase = ''
     runHook preCheck
     pushd dist
+    HOME=$TMPDIR
     ${python.interpreter} -c 'import qutip.testing as qt; qt.run()'
     popd
     runHook postCheck
```

**Comment by SuperSandro2000** [comment]
@mcncm ping

**Comment by FabianGD** [comment]
Closing this, as a newer version of the package is provided by #144102

**Pull request closed without merging**
